### PR TITLE
tests: Fix a bunch of filechecks

### PR DESCRIPTION
**Pull request by AntonLydike**
This fixes a bunch of "broken" filecheck checks.

Classes of mistakes fixed:

- Replace greedy captures for SSA value names in partial matches with `{{\S+}}` (match some non-whitespace chars), this was a problem, as the `{{.*}}` would greedily consume the whole line, leaving nothing to match for the next `CHECK-SAME` statement to match on.
- Replace expressions that "look" like variable capture (string substitution) `[[`-enclosed strings are reserved for [string substitution blocks](https://llvm.org/docs/CommandGuide/FileCheck.html#filecheck-string-substitution-blocks)
  - If one wants to match `[[` or `{{`, it's suggested to use the [literal modifier](https://llvm.org/docs/CommandGuide/FileCheck.html#directive-modifiers) (`CHECK{LITERAL}:`)
- Just plain old wrong checks ([func_and_arith_to_riscv_asm_flow.mlir](https://github.com/xdslproject/xdsl/compare/anton/fix-file-checks?expand=1#diff-40c6a572cd801477c4a2970789ad4c1e712548ba2d22321ef1a9b47c90feb3c1))
- `CHECK-NEXT`s without a previous `CHECK` line
- Removing a single, sad `CHECK-DAG` that was ignored by current filecheck anyway

### CODE REVIEW

File: tests/filecheck/projects/riscv-backend-paper/fill.mlir
Comment (PapyChacal) [inline tests/filecheck/projects/riscv-backend-paper/fill.mlir:32]: 🤯🤯

File: tests/filecheck/dialects/memref/memref_ops.mlir
Comment (PapyChacal) [inline tests/filecheck/dialects/memref/memref_ops.mlir:85]: Isn't there a way to just escape the `[[`?
Same question for all the `[[` -> `[`

File: tests/filecheck/dialects/gpu/ops.mlir
Comment (PapyChacal) [inline tests/filecheck/dialects/gpu/ops.mlir:145]: (This one is just OOC, I don't mind as this doesn't influence anything actually checked, so please ignore if you don't have time for it)

Those didn't fail, did they ?

File: tests/filecheck/backend/riscv/prologue_epilogue_insertion.mlir
Comment (PapyChacal) [inline tests/filecheck/backend/riscv/prologue_epilogue_insertion.mlir:81]: ❤️

File: tests/filecheck/backend/riscv/prologue_epilogue_insertion.mlir
Comment (PapyChacal) [inline tests/filecheck/backend/riscv/prologue_epilogue_insertion.mlir:33]: This one I can't wrap my head around; how did it not fail before ? Are CHECK-SAME handled at all by our filecheck?

File: tests/filecheck/backend/riscv/func_and_arith_to_riscv_asm_flow.mlir
Comment (PapyChacal) [inline tests/filecheck/backend/riscv/func_and_arith_to_riscv_asm_flow.mlir:16]: ❤️

**Review by PapyChacal — COMMENTED** [review]
Some interrogations (and just ❤️s, really meant those here ahah)

(For external readers, in this context, by "did not fail" I mean the previous state of the filecheck ran against a correct filecheck implementation)

File: tests/filecheck/projects/riscv-backend-paper/add.mlir
Comment (superlopuh) [inline tests/filecheck/projects/riscv-backend-paper/add.mlir:45]: hmm

File: tests/filecheck/projects/riscv-backend-paper/fill.mlir
Comment (superlopuh) [reply tests/filecheck/projects/riscv-backend-paper/fill.mlir:32]: Yeah this is an improvement

File: tests/filecheck/projects/riscv-backend-paper/add.mlir
Comment (superlopuh) [reply tests/filecheck/projects/riscv-backend-paper/add.mlir:45]: ok maybe CHECK-SAME is the way here also, but the outer [] is actually significant

File: tests/filecheck/projects/riscv-backend-paper/pres.mlir
Comment (superlopuh) [inline tests/filecheck/projects/riscv-backend-paper/pres.mlir:83]: 👍

File: tests/filecheck/dialects/memref/memref_ops.mlir
Comment (AntonLydike) [reply tests/filecheck/dialects/memref/memref_ops.mlir:85]: Yes, there is! `FileCheck` supports `CHECK{LITERAL}: now I can use {{ and [[ however I want, as this will be interpreted as a literal string!`

File: tests/filecheck/dialects/gpu/ops.mlir
Comment (AntonLydike) [reply tests/filecheck/dialects/gpu/ops.mlir:145]: Nope, the current `filecheck` happily passed this check as good!

File: tests/filecheck/projects/riscv-backend-paper/add.mlir
Comment (AntonLydike) [reply tests/filecheck/projects/riscv-backend-paper/add.mlir:45]: I'll give it another try then!

**Review by webmiche — COMMENTED** [review]
The lords work! :heart:

**Review by georgebisbas — APPROVED** [review]

File: tests/filecheck/dialects/memref/memref_ops.mlir
Comment (PapyChacal) [reply tests/filecheck/dialects/memref/memref_ops.mlir:85]: So you have to choose between escaping those or being able to use any regex and substitution?

File: tests/filecheck/projects/riscv-backend-paper/add.mlir
Comment (superlopuh) [inline tests/filecheck/projects/riscv-backend-paper/add.mlir:45]: 😵‍💫

**Review by superlopuh — APPROVED** [review]
I think this is ok but the `[[` situation is a little worrying, since it's so common in nd arrays. Not sure what the best way to fix that would be.

**Comment by AntonLydike** [comment]
> I think this is ok but the `[[` situation is a little worrying, since it's so common in nd arrays. Not sure what the best way to fix that would be.

This will be possible as soon as we switch to the new filecheck, as that supports `CHECK{LITERAL}:`, which makes no attempts at parsing regexes or string substitutions

File: tests/filecheck/backend/riscv/prologue_epilogue_insertion.mlir
Comment (AntonLydike) [reply tests/filecheck/backend/riscv/prologue_epilogue_insertion.mlir:33]: Can confirm `CHECK-SAME` is not handled by old filecheck. Would only merge this PR when we are immediately ready to merge the switch to new filecheck.

File: tests/filecheck/transforms/cse.mlir
Comment (AntonLydike) [inline tests/filecheck/transforms/cse.mlir:26]: These were duplicated with the (previously ignored `CHECK-LABEL`s above.

File: tests/filecheck/transforms/cse.mlir
Comment (AntonLydike) [inline tests/filecheck/transforms/cse.mlir:133]: `@down_propagate` was not a unique label, (FileCheck requires labels to be globally unique), so I made a change here to make it unique.

File: tests/filecheck/dialects/memref/memref_ops.mlir
Comment (AntonLydike) [reply tests/filecheck/dialects/memref/memref_ops.mlir:85]: In combination with `CHECK-SAME` this becomes almost bearable I think. See https://github.com/xdslproject/xdsl/pull/2828/files#diff-19f832aba596f43147f16a174b81a972eff70321b597de54501ea42b60cccb15

Comment (PapyChacal) [reply tests/filecheck/dialects/memref/memref_ops.mlir:85]: Yeah okay, that's a solution!

File: tests/filecheck/projects/riscv-backend-paper/add.mlir
Comment (AntonLydike) [reply tests/filecheck/projects/riscv-backend-paper/add.mlir]: Removed!

**Comment by AntonLydike** [comment]
I resolved the `[{{}}[` situation!

**Comment by AntonLydike** [comment]
I want to merge this one as close together with #2828 as possible, so I'll wait until that one has got the reviews necessary as well.

**Comment by tobiasgrosser** [comment]
Why is there a need to merge them close together.if this breaks the build,they should be in the same pr,no?

**Comment by AntonLydike** [comment]
> Why is there a need to merge them close together.if this breaks the build,they should be in the same pr,no?

You are right, I'll just merge this now

**Pull request merged**
